Kickstarter-iOS/Features/PledgePaymentMethods/Controller/PledgePaymentMethodsViewController.swift

```diff
@@ -223,6 +222,98 @@ final class PledgePaymentMethodsViewController: UIViewController {
     }
   }
 
+  private func didConfirmAddedPaymentOption(
+    paymentOption: PaymentSheet.FlowController.PaymentOptionDisplayData,
+    forClientSecret clientSecret: String,
+    forPaymentSheetType paymentSheetType: PledgePaymentSheetType
+  ) {
+    // Stripe also defines this constant, but it's defined internally to the Stripe SDK :(
+    let linkIdentifier = "link"
+
+    // For regular payment types, continue with the display information they gave us
+    if paymentOption.paymentMethodType != linkIdentifier {
+      let paymentDisplayData = PaymentSheetPaymentOptionsDisplayData(
+        image: paymentOption.image,
+        label: paymentOption.label
+      )
+
+      self.viewModel.inputs.paymentSheetDidAdd(newCard: paymentDisplayData, clientSecret: clientSecret)
+      return
+    }
+
+    // For link payment types, fetch the underlying payment method so we can get at the card type and label
+
+    switch paymentSheetType {
+    case .setupIntent:
+      STPAPIClient.shared
+        .retrieveSetupIntent(
+          withClientSecret: clientSecret,
+          expand: ["payment_method"]
+        ) { [weak self] intent, _ in
+          guard let strongSelf = self else { return }
+          guard let paymentDisplayData = strongSelf.paymentDisplayData(forLink: intent?.paymentMethod)
+          else { return }
+          strongSelf.viewModel.inputs.paymentSheetDidAdd(
+            newCard: paymentDisplayData,
+            clientSecret: clientSecret
+          )
+        }
+
+    case .paymentIntent:
+      STPAPIClient.shared
+        .retrievePaymentIntent(
+          withClientSecret: clientSecret,
+          expand: ["payment_method"]
+        ) { [weak self] intent, _ in
+          guard let strongSelf = self else { return }
+          guard let paymentDisplayData = strongSelf.paymentDisplayData(forLink: intent?.paymentMethod)
+          else { return }
+          strongSelf.viewModel.inputs.paymentSheetDidAdd(
+            newCard: paymentDisplayData,
+            clientSecret: clientSecret
+          )
+        }
+    }
+  }
+
+  private func paymentDisplayData(forLink optionalPaymentMethod: STPPaymentMethod?)
+    -> PaymentSheetPaymentOptionsDisplayData? {
+    guard let paymentMethod = optionalPaymentMethod else {
+      return nil
+    }
+
+    let formattedLabel = self.formatLinkLabel(paymentMethod.label)
+    return PaymentSheetPaymentOptionsDisplayData(
+      image: paymentMethod.image,
+      label: formattedLabel ?? paymentMethod.label
+    )
+  }
+
+  private func formatLinkLabel(_ label: String) -> String? {
+    // Link gives us a label like "Visa 1234"; reformat it to match our UI
+    do {
+      let regex = try NSRegularExpression(pattern: "\\d{4}")
```

Comment:
Also we may want this to be in a helper function so that the settings payment method view can use it as well.

Reply:
Haha I'll add a comment. It's just looking for 4 digits. As for the settings page, in this case it doesn't need it, because we never actually display/fetch the payment method there. On that page, it just stores the card immediately and we fetch the stored card from Kickstarter instead of from Stripe.